region/client.go

```diff
@@ -701,6 +715,7 @@ func (c *client) send(rpc hrpc.Call) (uint32, error) {
 	if err != nil {
 		return id, fmt.Errorf("failed to marshal request header: %s", err)
 	}
+	dropRequestHeader(header)
```

Comment:
Why `dropRequestHeader` here instead of deferring it to the end after `header := getRequestHeader`?
Wouldn't calling `dropRequestHeader` here interfere with the panic error message below (if it does ever happen, even though it shouldn't)?

Reply:
I want to return the header object for reuse as soon as possible, to make it more likely it will be reused/there will be fewer allocated at one time.
This is the earliest line where it can be dropped. If I used a defer we would continue to hold onto the header object through the point where we do the write on the socket, which may take a while to complete. The fact that we don't return it on error cases is not a big deal, it just means the object will be GC'd rather than reused.

That all said, I would like to pull out this code that builds up the message into its own function that is separate from the write on the socket. With that done I could use a defer to release the header object in a more reasonable place. And with that code pulled out into its own function we can write better tests and benchmarks for it.

Reply:
Refactored the code.